### ISSUE TITLE
protocol::wire::stream::Connection not usable as type inside a struct.

### ISSUE THREAD

**Issue by braincow** [braincow/guessnet#1]
Compiler error:
```
error[E0243]: wrong number of type arguments: expected at least 2, found 0
  --> src/network/server.rs:22:11
   |
22 |     conn: protocol::wire::stream::Connection,
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected at least 2 type arguments
```

Most likely a n00b error on Rust, but how to work around this? Code in "protocol" branch of the repo where intention is to replace simple text based protocol with something more deterministic: https://github.com/braincow/guessnet/blob/9d5c1ed932a723c23a5009b02901ec080418251b/src/network/server.rs#L22

**Comment by braincow** [comment]
@dylanmckay as a creator of protocol crate, any pointers?

**Comment by dylanmckay** [comment]
Hello @braincow 

It looks like you need to specify the generic types required by the [`Connection` type](https://docs.rs/protocol/0.3.6/protocol/wire/stream/struct.Connection.html) - namely `P: Parcel`, `S:  Read + Write`, and `M: protocol::wire::middleware::Pipeline`.

The actual type should be specified like 

```rust
struct Foo {
    conn: protocol::wire::stream::Connection<P, S, M>, 
}
```

For the protocol crate, a `Parcel` is anything that can be sent over the protocol. For example, in your case, I believe that this is your [`Packet` enum](https://github.com/braincow/guessnet/blob/9d5c1ed932a723c23a5009b02901ec080418251b/src/network/protocol.rs).

`S` is the underlying stream type that you are working with. For example, if this was a `TcpStream`, then `S` would be `TcpStream`. If you were reading from a file as a connection, `S = std::fs::File`. I think your example uses `TcpStream`.

`M` is just your desired middleware pipeline, which can be left as the default of `protocol::wire::middleware::pipeline::Default`. This type parameter has a [default](https://docs.rs/protocol/0.3.6/protocol/wire/middleware/pipeline/struct.Default.html) and so you don't have to specify it.

**tl;dr**

do this

```rust
use Packet; // from protocol.rs

struct Foo
    conn: protocol::wire::stream::Connection<Packet, TcpStream>, 
}

**Comment by braincow** [comment]
Thank you very much for the pointers @dylanmckay. At least rustc did not complain about that anymore. It did not occur to me for some reason to use Packet in there and I pretty much tried everything else (including Parcel).

**Comment by braincow** [comment]
Thanks again @dylanmckay for helping out. This is the final product: https://github.com/braincow/guessnet/releases/tag/0.2.0. It took some time actually as Rust n00b to reverse engineer how to receive packets. This was not evident from the examples included in the crate so I might have ended up doing something silly.
